File: Week_10/Practicum/Find if Path Exists in Graph.py
def dfs(graph, visited, current, target):
    if current not in visited or visited[current]:
        return

    if current == target:
        return True

    visited[current] = True

    for node in graph[current]:
        if dfs(graph, visited, node, target):
            return True

    return False


class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        if source == destination:
            return True

        graph = defaultdict(list)
        for edge in edges:
            graph[edge[0]].append(edge[1])
            graph[edge[1]].append(edge[0])

        visited = {node: False for node in graph.keys()}

        return dfs(graph, visited, source, destination)

File: Week_10/Practicum/Find if Path Exists in Graph.cpp
class Solution {
public:
    bool validPath(int n, vector<vector<int>>& edges, int source, int destination) {
        vector<vector<int>> graph(n);
        for (auto& edge : edges) {
            graph[edge[0]].push_back(edge[1]);
            graph[edge[1]].push_back(edge[0]);
        }

        vector<bool> visited(n, false);
        queue<int> q;
        q.push(source);
        visited[source] = true;

        while (!q.empty()) {
            int curr = q.front(); q.pop();
            if (curr == destination) return true;
            for (auto &nbr : graph[curr]) {
                if (!visited[nbr]) {
                    visited[nbr] = true;
                    q.push(nbr);
                }
            }
        }

        return false;
    }
};
